### PR TITLE
Add assembling into shaderc.

### PR DESCRIPTION
**Pull request by antiagainst**
Addresses #80.

TODO: more tests for command-line.

### CODE REVIEW

**Comment by antiagainst** [comment]
@httpdigest: I assume you'll be interested in this PR. :)

**Comment by httpdigest** [comment]
Nice, thanks for implementing! Will test when it's merged.

File: libshaderc_util/src/spirv_tools_wrapper.cc [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  *dest = os.str();
+}
+
+}  // anonymous namespace
+
+namespace shaderc_util {
+
+bool SpirvToolsDisassemble(const std::vector<uint32_t>& binary,
+                           std::string* text_or_error) {
+  auto spvtools_context = spvContextCreate(SPV_ENV_UNIVERSAL_1_0);
```
Comment (dekimir) [inline libshaderc_util/src/spirv_tools_wrapper.cc]: Might be better to use `SPV_ENV_VULKAN*` here.  Right now it doesn't matter, but it would be more future-proof.

File: libshaderc_util/src/spirv_tools_wrapper.cc [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+  spvDiagnosticDestroy(spvtools_diagnostic);
+  spvTextDestroy(disassembled_text);
+  spvContextDestroy(spvtools_context);
+
+  return result;
+}
+
+bool SpirvToolsAssemble(string_piece assembly, std::vector<uint32_t>* binary,
+                        std::string* errors) {
+  auto spvtools_context = spvContextCreate(SPV_ENV_UNIVERSAL_1_0);
```
Comment (dekimir) [inline libshaderc_util/src/spirv_tools_wrapper.cc]: `ENV_VULKAN`

File: libshaderc_util/src/spirv_tools_wrapper.cc [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  spv_binary assembled_binary = nullptr;
+  spv_diagnostic spvtools_diagnostic = nullptr;
+
+  const bool result =
+      spvTextToBinary(spvtools_context, assembly.data(), assembly.size(),
+                      &assembled_binary, &spvtools_diagnostic) == SPV_SUCCESS;
+  if (result) {
+    binary->assign(assembled_binary->code,
```
Comment (dekimir) [inline libshaderc_util/src/spirv_tools_wrapper.cc]: Boo, extra copy. ;)

File: libshaderc_util/src/spirv_tools_wrapper.cc (hunk 1 above)
Comment (antiagainst) [inline libshaderc_util/src/spirv_tools_wrapper.cc]: Done.

File: libshaderc_util/src/spirv_tools_wrapper.cc (hunk 2 above)
Comment (antiagainst) [inline libshaderc_util/src/spirv_tools_wrapper.cc]: Done.

File: libshaderc_util/src/spirv_tools_wrapper.cc (hunk 3 above)
Comment (antiagainst) [inline libshaderc_util/src/spirv_tools_wrapper.cc]: Good catch!

I've converted the code to use a class hierarchy for compilation results. Now we can delegate the destruction of spv_binary to the proper compilation result subclass.

File: libshaderc/src/shaderc_private.h [hunk 4]
```diff
@@ -42,6 +43,46 @@ struct shaderc_compilation_result {
       shaderc_compilation_status_null_result_object;
 };
 
+// Compilation result class using a vector for holding the compilation
+// output data.
+class shaderc_compilation_result_vector : public shaderc_compilation_result {
```
Comment (dekimir) [inline libshaderc/src/shaderc_private.h]: At this point, we should seriously consider having a third subclass that contains a text result.  It's weird to have two different compilation_results for SPIR-V words but only one compilation_result for two different outputs (text vs. SPIR-V).

File: glslc/test/assembly.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    glslc_args = [shader]
+    expected_object_filenames = ('a.spv',)
+
+
+@inside_glslc_testsuite('SpirvAssembly')
+class TestDashCAssemblyFile(expect.ValidObjectFile):
+    """Tests that -c works with SPIR-V assembly file."""
+
+    shader = FileShader(empty_main_assembly(), '.spvasm')
+    glslc_args = ['-c', shader]
```
Comment (dekimir) [inline glslc/test/assembly.py]: Consider adding tests with multiple shader-file arguments: several `.spvasm` files, one `.spvasm` and one `.frag`, etc.

File: glslc/test/glslc_test_framework.py [hunk 6]
```diff
@@ -147,7 +147,7 @@ def __init__(self, returncode, stdout, stderr, directory, input_filenames):
         self.stderr = stderr
         # temporary directory where the test runs
         self.directory = directory
-        # the names of input shader files (without path)
+        # the names of input shader files (may with path)
```
Comment (dekimir) [inline glslc/test/glslc_test_framework.py]: "may with path" doesn't parse.  Consider "potentially including paths" or something similar.

File: libshaderc/CMakeLists.txt [hunk 7]
```diff
@@ -16,7 +16,8 @@ find_package(Threads)
 target_link_libraries(shaderc PRIVATE
   glslang OSDependent OGLCompiler glslang ${CMAKE_THREAD_LIBS_INIT})
 target_link_libraries(shaderc PRIVATE shaderc_util)
-target_link_libraries(shaderc PRIVATE SPIRV)
+target_link_libraries(shaderc PRIVATE SPIRV)  # from glslang
+target_link_libraries(shaderc PRIVATE SPIRV-Tools)
```
Comment (dekimir) [inline libshaderc/CMakeLists.txt]: Hm, how did this work before??

Comment (antiagainst) [inline libshaderc/CMakeLists.txt]: Because previously we hide everything related to spirv-tools in `libshaderc_util`, but now we `#include`d `libspirv.h` in `shaderc_private.h`.

File: libshaderc/src/shaderc_private.h (hunk 4 above)
Comment (antiagainst) [inline libshaderc/src/shaderc_private.h]: Yeah, make sense for me. But I'd like to put that as a separate PR since this one is already fairly large.

Comment (dekimir) [inline libshaderc/src/shaderc_private.h]: OK.

File: glslc/test/glslc_test_framework.py (hunk 6 above)
Comment (antiagainst) [inline glslc/test/glslc_test_framework.py]: Done. See 016b4b4.

File: glslc/test/assembly.py (hunk 5 above)
Comment (antiagainst) [inline glslc/test/assembly.py]: Done. See 016b4b4.

File: glslc/test/assembly.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+class TestAssemblyFileWithOnlyComments(expect.ValidObjectFile):
+    """Tests that glslc accepts an assembly file with only comments inside."""
+
+    shader = FileShader(assembly_comments(), '.spvasm')
+    glslc_args = ['-c', shader]
+
+
+@inside_glslc_testsuite('SpirvAssembly')
+class TestEmptyAssemblyFile(expect.ValidObjectFile):
+    """Tests that glslc accepts an empty assembly file."""
```
Comment (antiagainst) [inline glslc/test/assembly.py]: clang accepts an empty `.s` file and generated a corresponding non-empty `.o` file.

File: glslc/test/assembly.py [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+    shader1 = FileShader(empty_main_assembly(), '.spvasm')
+    shader2 = FileShader(empty_main(), '.vert')
+    shader3 = FileShader(empty_main(), '.frag')
+    glslc_args = ['-c', shader1, shader2, shader3]
+
+
+@inside_glslc_testsuite('SpirvAssembly')
+class TestShaderStageWithAssemblyFile(expect.ErrorMessage):
+    """Tests that assembly files don't work with -fshader-stage"""
```
Comment (antiagainst) [inline glslc/test/assembly.py]: clang fails given `clang -x c++ some.s`.

File: glslc/src/file_compiler.cc [hunk 10]
```diff
@@ -54,10 +54,10 @@ bool FileCompiler::CompileShaderFile(const std::string& input_file,
     error_file_name = "<stdin>";
   }
 
-  string_piece source_string;
+  string_piece source_string = "";
   if (!input_data.empty()) {
-    source_string = string_piece(&(*input_data.begin()),
-                                 &(*input_data.begin()) + input_data.size());
+    source_string = {&(*input_data.begin()),
+                     &(*input_data.begin()) + input_data.size()};
```
Comment (dekimir) [inline glslc/src/file_compiler.cc]: Nit: could've used `front()` and `back()` here.

**Comment by dekimir** [comment]
+2

File: glslc/src/file_compiler.cc (hunk 10 above)
Comment (antiagainst) [inline glslc/src/file_compiler.cc]: Done.
